src/common/cockpitpipe.c

```diff
@@ -1554,3 +1652,58 @@ cockpit_pipe_get_environ (const gchar **input,
 
   return env;
 }
+
+static void
+on_throttle_pressure (GObject *object,
+                      gboolean throttle,
+                      gpointer user_data)
+{
+  CockpitPipe *self = COCKPIT_PIPE (user_data);
+  if (throttle)
+    {
+      if (self->priv->in_source != NULL)
+        {
+          g_debug ("%s: applying back pressure in pipe", self->priv->name);
+          stop_input (self);
+        }
+    }
+  else
+    {
+      if (self->priv->in_source == NULL && !self->priv->in_done)
+        {
+          g_debug ("%s: relieving back pressure in pipe", self->priv->name);
+          start_input (self);
+        }
+    }
+}
+
+/**
+ * cockpit_pipe_throttle:
+ * @self: The pipe
+ * @pressure: An object to throttle this channel
+ *
+ * Have this pipe throttle its inward flow based on the "pressure"
+ * signal from the @pressure object.
+ */
+void
+cockpit_pipe_throttle (CockpitPipe *self,
+                       GObject *pressure)
+{
+  g_return_if_fail (COCKPIT_IS_PIPE (self));
+
+  if (self->priv->pressure)
+    {
+      g_signal_handler_disconnect (self->priv->pressure, self->priv->pressure_sig);
+      g_object_remove_weak_pointer (G_OBJECT (self->priv->pressure), (gpointer *)&self->priv->pressure);
+      self->priv->pressure = NULL;
+    }
+
+  self->priv->pressure_sig = 0;
```

Comment:
dubious.  In case the object is removed via weak pointer notification, pressure_sig will contain junk.  Your code seems safe against that possibility (ie: it always only checks ->pressure != NULL) so why do this unnecessary and confusing cleanup here?